### PR TITLE
We need to sync gradients with regards to input in MQA

### LINKED ISSUE #35: Unstable training with MQA and TP>1
Models trained with MQA and TP>1 have exploding loss after a few thousand steps.
This issue does not happen when training the same models with MHA, or with TP=1.

image (1)

### PR DESCRIPTION
**Pull request by thomasw21**
Fixes: #35 

The issue is that in MQA we need to sync gradients with regards to weigths/bias as they are duplicated over TP. However we also need to synchronize gradient with regards to input, similarly to how it's done in ColumnLinear

### CODE REVIEW

File: megatron/optimizer/optimizer.py [hunk 1]
```diff
@@ -272,23 +272,27 @@ def allreduce_key_value_grads(self, args):
...
-        if len(grads)>0:
-            coalesced = _flatten_dense_tensors(grads)
-            torch.distributed.all_reduce(
-                coalesced, group=mpu.get_tensor_model_parallel_group())
-            for buf, synced in zip(grads, _unflatten_dense_tensors(
-                    coalesced, grads)):
-                buf.copy_(synced)
+        # All-reduce kv parameters across model parallel nodes
+        # when sequence parallelism is used
+        if mpu.get_tensor_model_parallel_world_size() > 1 and \
```
Comment (jlamypoirier) [inline megatron/optimizer/optimizer.py:277]: Can you please move the conditions outside this function and put it with the checks for mqa? (https://github.com/bigcode-project/Megatron-LM/pull/37/files#diff-4125b0b128f265b3cdd9786f9419957a9bb4c75a93aba7e17fb0ba3fed1da199L545, https://github.com/bigcode-project/Megatron-LM/pull/37/files#diff-dabd5ffec3ce00acedf6bf253e236186da42c61e2501d7d3f5a84645bfb6578bL337)
This would reduce the diff (the rest of this function is unchanged right?) and skip some unnecessary synchronization with the timers.

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -611,12 +611,16 @@ def forward(self, hidden_states, attention_mask,
              value_layer) = mpu.split_tensor_along_last_dim(mixed_x_layer, 3)
         elif self.attention_type == AttnType.self_attn and self.attention_head_type == 'multiquery':
             kv_input=hidden_states
-            if get_args().sequence_parallel:
-                # The linear layer doesn't gather the sequence-parallel.
-                kv_input = mpu.gather_from_sequence_parallel_region(kv_input, tensor_parallel_output_grad=False)
+
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:614]: Unnecessary diff

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -611,12 +611,16 @@ def forward(self, hidden_states, attention_mask,
...
             # Attention heads [sq, b, h] --> [sq, b, (2 * hn)]
             mixed_kv_layer = self.key_value(kv_input)
 
+            if get_args().sequence_parallel:
+                # The linear layer doesn't gather the sequence-parallel.
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:619]: Can you please improve/update this comment?

**Review by jlamypoirier — COMMENTED** [review]
Some minor suggestions, otherwise the PR looks great!

**Comment by thomasw21** [comment]
Closing in favor or #39
